docs/caching.md

```diff
@@ -0,0 +1,281 @@
+
+
+## Example 5: Use the global cache to speed up your app for all users
+
+Going back to our original function, let's add a widget to control the value of `b`:
+
+```python
+import streamlit as st
+import time
+
+@st.cache(suppress_st_warning=True)
+def expensive_computation(a, b):
+    st.write("Cache miss: expensive_computation(", a, ",", b, ") ran")
+    time.sleep(2)  # This makes the function take 2s to run
+    return a * b
+
+a = 2
+b = st.slider("Pick a number", 0, 10)  # 👈 Change this
+res = expensive_computation(a, b)
+
+st.write("Result:", res)
+```
+
+What you'll see:  
+
+* If you move the slider to a number Streamlit hasn't seen before, you'll have a cache miss again. And every subsequent rerun with the same number will be a cache hit, of course.
+* If you move the slider back to a number Streamlit has seen before, the cache is hit and the app is fast as expected.
+
+In computer science terms, what is happening here is that [`@st.cache`](api.html#streamlit.cache) is [memoizing](https://en.wikipedia.org/wiki/Memoization) `expensive_computation(a, b)`.
+
+But now let's go one step further! Try the following:
+
+1. Move the slider to a number you haven't tried before, such as 9.
+2. Pretend you're another user by opening another browser tab pointing to your Streamlit app (usually at http://localhost:8501)
+3. In the new tab, move the slider to 9.
+
+Notice how this is actually a cache hit! That is, you don't actually see the "Cache miss" text on the second tab even though that second user never moved the slider to 9 at any point prior to this.
+
+This happens because the Streamlit cache is global to all users. So everyone contributes to everyone else's performance.
+
+## Example 6: Mutating cached values
+
+As mentioned in the How Caching Works section, the Streamlit cache stores items by reference. This allows the Streamlit cache to support structures that aren't memory-managed by Python, such as TensorFlow objects. However, it can also lead to unexpected behavior — which is why Streamlit has a few checks to guide developers in the right direction. Let's look into those checks now.
+
+Let's write an app that has a cached function which returns a mutable object, and then let's follow up by mutating that object:
+
+```python
+import streamlit as st
+import time
+
+@st.cache(suppress_st_warning=True)
+def expensive_computation(a, b):
+    st.write("Cache miss: expensive_computation(", a, ",", b, ") ran")
+    time.sleep(2)  # This makes the function take 2s to run
+    return {"output": a * b}  # 👈 Mutable object
+
+a = 2
+b = 21
+res = expensive_computation(a, b)
+
+st.write("Result:", res)
+
+res["output"] = "result was manually mutated"  # 👈 Mutate cached value
+
+st.write("Mutated result:", res)
+```
+
+When you run this app for the first time, you should see three messages on the screen:
+
+* Cache miss (...)
+* Result: {output: 42}
+* Mutated result: {output: "result was manually mutated"}
+
+So what's up?
```

Comment:
You're missing this text above "So what's up?":

> No surprises here. But now notice what happens when you rerun you app (i.e. press R):
>
>     Result: {output: "result was manually mutated"}
>     Mutated result: {output: "result was manually mutated"}
>     <Warning> Cached object mutated. (...)

Reply:
Fixed.